### PR TITLE
Implemented most important parts of ISteamMatchmaking

### PR DESCRIPTION
**Pull request by kvittokonito**
All new methods and events personally tested and confirmed working.

### CODE REVIEW

**Comment by hokein** [comment]
@aftokinito Many thanks for the contribution, this is **awesome**!

The patch looks good in general:
- could you please update the PR description to summarize the changes of this PR? As far as I understand, you implemented SteamMatchmaking APIs; added some friends APIs; exposed a RunCallback APIs;
- could you please also add API documentation in https://github.com/greenheartgames/greenworks/tree/master/docs? 
-  for the `runCallbacks` API, I'm not sure we should expose it to clients -- as we have already called it periodically in Greenworks, so clients don't need to call it; exposing it will make client implement another event loop; it seems this is likely an electron issue, have you tested on different electron version to confirm?

Again, thanks for you time!

**Comment by TheRealDannyyy** [comment]
@kvittokonito Huge fan of your work, especially adding the support for Steam rich presence!

_**Hokein**_ said that he would be up for [building a new release](https://github.com/greenheartgames/greenworks/issues/234) soon. I would really appreciate if you could finish and merge this PR soon, so that it can be included in the next release.

_Cheers!_

**Comment by kvittokonito** [comment]
I will try and book some time to write the documentation.  
My company does no longer use GreenWorks as we've moved away from NodeJS so I cannot really make any promises regarding when I will have time to write the requested documentation.

**Comment by TheRealDannyyy** [comment]
Thanks for the response!

Perhaps this could be merged into the next build regardless and the docs could be added later. 
@hokein would you agree with this approach?

**Comment by kvittokonito** [comment]
All the stuff added by this PR should be working fine provided nothing in master has broken it since it was opened, all the features I added are used in the old Electron version of our game. Most functionality should also be basically a mirror of the SteamWorks API.

**Comment by ixevix** [comment]
So what's the status on getting this merged? I tried to fork the repo but I the pull request isn't available in the forked one. Is there any way to add it to my fork?

**Comment by ixevix** [comment]
Ah, I'm stupid, I can just use your fork instead.

**Comment by ixevix** [comment]
Hmm, seems I can't get it to build. I'm building against Steamworks SDK 1.41 like the documentation says and I get these errors on two different linux hosts:

```
In file included from ../src/api/steam_api_settings.cc:17:
../src/steam_event.h: At global scope:
../src/steam_event.h:37:8: error: extra qualification ‘greenworks::SteamEvent::’ on member ‘OnLobbyCreated’ [-fpermissive]
   void SteamEvent::OnLobbyCreated(int status_code, uint64 SteamIdLobby);
        ^~~~~~~~~~
../src/steam_event.h:38:8: error: extra qualification ‘greenworks::SteamEvent::’ on member ‘OnLobbyDataUpdate’ [-fpermissive]
   void SteamEvent::OnLobbyDataUpdate(uint64 SteamIdLobby, uint64 SteamIdMember, bool Success);
        ^~~~~~~~~~
../src/steam_event.h:39:8: error: extra qualification ‘greenworks::SteamEvent::’ on member ‘OnLobbyEnter’ [-fpermissive]
   void SteamEvent::OnLobbyEnter(uint64 SteamIdLobby, int ChatPermissions, bool Locked, int ChatRoomEnterResponse);
        ^~~~~~~~~~
../src/steam_event.h:40:8: error: extra qualification ‘greenworks::SteamEvent::’ on member ‘OnLobbyInvite’ [-fpermissive]
   void SteamEvent::OnLobbyInvite(uint64 SteamIdUser, uint64 SteamIdLobby, uint64 GameId);
        ^~~~~~~~~~
../src/steam_event.h:41:8: error: extra qualification ‘greenworks::SteamEvent::’ on member ‘OnGameLobbyJoinRequested’ [-fpermissive]
   void SteamEvent::OnGameLobbyJoinRequested(uint64 SteamIdLobby, uint64 SteamIdUser);
        ^~~~~~~~~~

```

**Comment by ixevix** [comment]
@kvittokonito anything?

**Comment by Karawasa** [comment]
@kvittokonito @TheRealDannyyy @hokein 

I'm having the same build issue. Can one of you guys help out? Matchmaking is an important aspect and I'd love to have this as part of greenworks.

**Comment by mastastealth** [comment]
@ixevix Have you tried 1.42 for the SDK? That's the version in the main README at least.

**Comment by ixevix** [comment]
Actually I tried building against all the following versions: 1.38a, 1.39, 1.40, 1.41, 1.42, 1.45 and 1.46.

**Comment by TheRealDannyyy** [comment]
Any progress on this amazing PR?
Let's please get this out and running soon.

**Comment by ixevix** [comment]
I figured out how to build it. Just remove the "greenworks::SteamEvent::" part to match the other declarations in src/steam_event.h and run npm install. At least it compiled for me.

File: src/steam_event.h [hunk 1]
```diff
@@ -34,6 +34,12 @@ class SteamEvent : public greenworks::SteamClient::Observer {
   void OnMicroTxnAuthorizationResponse(uint32 AppID,
                                        uint64 OrderID,
                                        bool Autorized) override;
+  void SteamEvent::OnLobbyCreated(int status_code, uint64 SteamIdLobby);
```
Comment (ixevix) [inline src/steam_event.h]: Remove "SteamEvent::" to get it to compile

**Comment by TheRealDannyyy** [comment]
@ixevix Do you think you could fix this up and get it ready for a new build?

**Comment by ixevix** [comment]
I just told you how you can do it yourself. Just clone the kvittokonito fork of this repo and make the changes I posted. I'm not maintaining this repo so I cannot make a build for you.

File: src/steam_event.h (hunk 1 above)
Comment (kvittokonito) [reply src/steam_event.h]: Fixed.

**Comment by TheRealDannyyy** [comment]
Looks like we got some conflicts after the recent commits. Would appreciate a fix!

**Pull request closed without merging**